core/chains/evm/gas/universal_estimator.go

```diff
@@ -0,0 +1,412 @@
+}
+
+func NewUniversalEstimator(lggr logger.Logger, client universalEstimatorClient, cfg UniversalEstimatorConfig, chainID *big.Int, l1Oracle rollups.L1Oracle) *UniversalEstimator {
+	return &UniversalEstimator{
+		client:   client,
+		logger:   logger.Named(lggr, "UniversalEstimator"),
+		config:   cfg,
+		chainID:  chainID,
+		l1Oracle: l1Oracle,
+		wg:       new(sync.WaitGroup),
+		stopCh:   make(chan struct{}),
+	}
+}
+
+func (u *UniversalEstimator) Start(context.Context) error {
+	return u.StartOnce("UniversalEstimator", func() error {
+		if u.config.BumpPercent < MinimumBumpPercentage {
+			return fmt.Errorf("BumpPercent: %s is less than minimum allowed percentage: %s",
+				strconv.FormatUint(uint64(u.config.BumpPercent), 10), strconv.Itoa(MinimumBumpPercentage))
+		}
+		if u.config.EIP1559 && u.config.RewardPercentile > ConnectivityPercentile {
+			return fmt.Errorf("RewardPercentile: %s is greater than maximum allowed connectivity percentage: %s",
+				strconv.FormatUint(uint64(u.config.RewardPercentile), 10), strconv.Itoa(ConnectivityPercentile))
+		}
+		if u.config.EIP1559 && u.config.BlockHistorySize == 0 {
+			return fmt.Errorf("BlockHistorySize is set to 0 and EIP1559 is enabled")
+		}
+		u.wg.Add(1)
+		go u.run()
+
+		return nil
+	})
+}
+
+func (u *UniversalEstimator) Close() error {
+	return u.StopOnce("UniversalEstimator", func() error {
+		close(u.stopCh)
+		u.wg.Wait()
+		return nil
+	})
+}
+
+func (u *UniversalEstimator) run() {
+	defer u.wg.Done()
+
+	t := services.NewTicker(u.config.CacheTimeout)
+	for {
+		select {
+		case <-u.stopCh:
+			return
+		case <-t.C:
+			if u.config.EIP1559 {
+				if _, err := u.FetchDynamicPrice(); err != nil {
+					u.logger.Error(err)
+				}
+			} else {
+				if _, err := u.FetchGasPrice(); err != nil {
+					u.logger.Error(err)
+				}
+			}
+		}
+	}
+}
+
+// GetLegacyGas will fetch the cached gas price value.
+func (u *UniversalEstimator) GetLegacyGas(ctx context.Context, _ []byte, gasLimit uint64, maxPrice *assets.Wei, opts ...feetypes.Opt) (gasPrice *assets.Wei, chainSpecificGasLimit uint64, err error) {
+	chainSpecificGasLimit = gasLimit
+	if gasPrice, err = u.getGasPrice(); err != nil {
+		return
+	}
+
+	if gasPrice.Cmp(maxPrice) > 0 {
+		u.logger.Warnf("estimated gas price: %s is greater than the maximum gas price configured: %s, returning the maximum price instead.", gasPrice, maxPrice)
+		return maxPrice, chainSpecificGasLimit, nil
+	}
+	return
+}
+
+// FetchGasPrice will use eth_gasPrice to fetch and cache the latest gas price from the RPC.
+func (u *UniversalEstimator) FetchGasPrice() (*assets.Wei, error) {
+	ctx, cancel := u.stopCh.CtxCancel(evmclient.ContextWithDefaultTimeout())
+	defer cancel()
+
+	gasPrice, err := u.client.SuggestGasPrice(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch gas price: %s", err)
+	}
+
+	promUniversalEstimatorGasPrice.WithLabelValues(u.chainID.String()).Set(float64(gasPrice.Int64()))
+
+	gasPriceWei := (*assets.Wei)(gasPrice)
```

Comment:
nit: this is probably ok but feel like this looks better
```suggestion
	gasPriceWei := assets.NewWei(gasPrice)
```